3.0.0 release notes && release notes automation

shilman opened the pull request:
Issue: #1046 

## What I did
- Release note proposed structure
- Rough draft of release notes

Inspired by:
- https://github.com/facebook/react/releases
- https://atom.io/releases

## What I need
- Other notable changes
- Discussion on release structure
- Suggestion for how to get a formatted list of PRs since last tagged release? => this will get added to the CHANGELOG as well as the release `tag`

## How to test

Browse the formatted release notes [HERE](https://github.com/shilman/storybook/blob/1046-3.0-release-notes/CHANGELOG.md)

shilman committed:
Merge remote-tracking branch 'storybooks/master'

shilman committed:
First draft of release notes

aaronmcadam commented:
> Suggestion for how to get a formatted list of PRs since last tagged release?

I use [Lerna's conventional commits feature](https://github.com/lerna/lerna#--conventional-commits), it works very nicely.

UsulPro commented:
> Suggestion for how to get a formatted list of PRs since last tagged release?

or maybe like this?

https://github.com/storybooks/storybook/pulls?page=1&q=is%3Apr+is%3Aclosed+merged%3A2017-01-15..2017-05-18&utf8=%E2%9C%93

apexskier commented on CHANGELOG.md:
```diff
@@ -1,4 +1,18 @@
+
+Storybook 3.0 is our first fully community-driven release! Notable changes:
+- Move from `@kadira` to `@storybooks` org across [github](https://github.com/storybooks/storybook/), [npm](FIXME), [docs](https://storybooks.js.org/)
+- Upgrade to Webpack2 [PR637](https://github.com/storybooks/storybook/pull/637)
```
This is a bit nitpicky, and really doesn't matter, but you can just include the URL and github will shorten it to a nicer looking link.

Example:
https://github.com/storybooks/storybook/pull/637
https://github.com/storybooks/storybook-addon-graphql/pull/19

```
https://github.com/storybooks/storybook/pull/637
https://github.com/storybooks/storybook-addon-graphql/pull/19
```

I think it's a slightly nicer format when viewing issues on github, and it falls back to the correct link when viewing code outside of github.

I'm totally fine using the `[PRXXX](link)` format if that's the desired format, but since it looks kind of inconsistent in the changelog's history, I thought I'd bring it up.

shilman replied:
@apexskier I just used the  PR format since it was already setup that way. I have converted to using normal links, although they don't seem to be collapsing properly (perhaps because I'm viewing them on a fork of the main repo???) Further suggestions much appreciated!

apexskier replied:
I _think_ the collapsing will be fixed once merged 😬

shilman commented:
@aaronmcadam conventional-commits works great. but it also requires that we get everybody to adopt that style including PRs from community members.

Somebody also suggested https://github.com/zeit/release, where the person making the release can manually classify each commit. But I guess it does not know about monorepos.

At any rate, it seems like best practices seems to be logging at the commit level. Does this mean we should get smarter about squashing commits and other more serious git usage?

cc @ndelangen @tmeasday @UsulPro

ndelangen commented:
Can we link PR's merged to issues and list those in the CHANGELOG?

tmeasday commented:
I think systems that are based on "hygenic" commits are going to be hard to enforce on a community project unless we want to do a lot of PR rebasing, which sounds like a pain.

I would suggest we follow the process we do for Meteor and Apollo, which is to have a `vNEXT` section of the changelog, ask people to make PRs against a `dev` branch and include changes to the changelog in their PR (alongside doc changes too if possible). 

So a contributor would add a line to the `vNEXT` section of the changelog. Then the release process is to merge `dev` to `master`, re-label `vNEXT` as `vX.Y.Z`, and publish.

If we don't want to use `dev` we could do a different strategy with the same idea.

shilman commented:
@ndelangen @tmeasday I did a quick experiment of trying to pull out merged PR's that specifically referenced commits in the first line of their body. The rationale being that enforcing hygienic commits is probably too much work, but maybe we can do a better job with PRs. Also Github makes it easy for us to edit PR titles to make them more descriptive.

Here is the output of that experiment:
```
#1043 Update README after repo-restructure (#1042)
#1038 Deprecate builtin addons (#1017)
#1025 Deprecate built-in addons (#1017)
#1024 Updated issue triage guidelines (#733)
#1021 Add addonlinks repo (#1016)
#996 refactor: change NPM organisation from kadira to storybook (#773)
#989 Add typescript definition to addon-notes (#953)
#961 #959 add a max-height and center element with alignItems: center (#959)
```
Let me know if you think this ideas merit. If so I can pursue it further. If not, I'm tempted to go with @tmeasday 's suggestion of a manually-curated changelog, but enforced in PRs.

tmeasday commented:
> referenced commits

Do you mean "referenced issues"?

I can get behind the idea of updating the PR title instead of changing a changelog file. If it's reliable it's a great middle ground.

shilman commented:
@tmeasday Sorry, yes referenced issue. I updated my comment. 

OK, that's encouraging. I'll take a more complete pass at this today and we can at least give it a shot. If it turns out to be a pain, we can revert to a manual method later. Thanks!

shilman committed:
Add git URL to package.json for changelog generator

shilman commented:
This is the format generated by https://github.com/ahmdigital/github-publish-release:

```
- Updated issue triage guidelines #1024 (by @shilman)
- Updated dead getstorybook.io links to storybooks.js.org #988 (by @zjeraar)
- Update README after repo-restructure #1043 (by @shilman)
- Update links in PRs section of CONTRIBUTING.md #992 (by @UsulPro)
- Update jest to the latest version 🚀 #998 (by @greenkeeper[bot])
- Switch back to non-fork of react-inspector #1026 (by @ndelangen)
- Remove text transform uppercase for knob labels #991 (by @cnandreu)
- refactor: change NPM organisation from kadira to storybook #996 (by @aaronmcadam)
- MOVE & RENAME and storyshots && CHANGE “React Storybook” > “Storybook” #1044 (by @ndelangen)
- Fixing link #974 (by @she-dev)
- Fixes for types for knobs `number` #1001 (by @marcfallows)
- Fix typo in README #977 (by @rixth)
- Fix linting in addon-info #980 (by @UsulPro)
- First pass at update the contribution instructions #982 (by @tmeasday)
- Do not ignore story.js in prepublish #973 (by @subpopular)
- Deprecate builtin addons #1038 (by @shilman)
- Deprecate built-in addons #1025 (by @shilman)
- Dependency updates #1008 (by @ndelangen)
- CHANGE folder structure && CHANGE package-names #1031 (by @ndelangen)
- addon-info: make the info overlay be fixed  #914 (by @Kiwka)
- Add typescript definition to addon-notes #989 (by @fabien0102)
- Add propTablesExclude option #924 (by @elisherer)
- Add addonlinks repo #1021 (by @ndelangen)
```

shilman commented:
Here's the latest, joining PRs and issues per @ndelangen 's suggestion.

<details>
<summary>
### CLOSED issues with PRs
</summary>
- #404 Images gets extra height when displayed in storybook #974
- #733 Issue triage guidelines #1024
- #773 Change npm organisation from kadira to storybook #996
- #959 [BUG] Style: Title of Panels #961
- #1016 Restore storybook-addon-links #1021
- #1017 Deprecate addon-links and addon-actions as "built-in addons" #1038,#1025
- #1042 Update README #1043
</details>
<details>
<summary>
### OPEN issues with PRs
</summary>
- #953 Typescript compatibility #989
</details>
<details>
<summary>
### PRs without issues
</summary>
- #1044 MOVE & RENAME and storyshots && CHANGE “React Storybook” > “Storybook”
- #1026 Switch back to non-fork of react-inspector
- #1008 Dependency updates
- #1001 Fixes for types for knobs `number`
- #998 Update jest to the latest version 🚀
- #997 Update eslint-plugin-jest to the latest version 🚀
- #992 Update links in PRs section of CONTRIBUTING.md
- #991 Remove text transform uppercase for knob labels
- #988 Updated dead getstorybook.io links to storybooks.js.org
- #982 First pass at update the contribution instructions
- #980 Fix linting in addon-info
- #977 Fix typo in README
- #975 support the react 16 alpha for react-native lib
- #973 Do not ignore story.js in prepublish
- #972 FIX for test-cra was breaking the build
- #969 Update lerna to the latest version 🚀
- #967 Fix error when component has object as props
- #965 FIX webpack overriding && CLEANUP && ADD example
- #956 CHANGE to use react-split-view 0.1.63 over the fork
- #955 Add splitpane repo
- #929 add description field from __docgenInfo for prop table for info plugin
</details>

UsulPro commented on ROADMAP.md:
```diff
@@ -2,13 +2,10 @@
 * Automatic story generation (and edge case detection) based on propTypes.
 * Angular Support
 * Vue Support
-* UI addons (Add different panels like Action Logger)
```
> -* A clear guide to hack React Storybook

I think we need the new one since we've changed everything in the repo.

> -* Addon API and addons
-* React Native Support
-* UI addons (Add different panels like Action Logger)

Do we have any place to move it instead of just delete? Could it be cool to show wich roadmap points already done? What do you think about it, maybe right here one more section with the title 'realized steps' or checkboxes?

shilman replied:
I moved it to the root directory, not delete.

UsulPro replied:
I mean don't delete this points

shilman committed:
Merge remote-tracking branch 'storybooks/master' into 1046-3.0-release-notes

# Conflicts:
#	ROADMAP.md

shilman committed:
Update CHANGELOG according to feedback

- use plain links to PRs
- group closed PRs by issue type

ndelangen commented:
I think we're on the right track.

If we could do things like extract labels from issues and do something with that in the CHANGELOG, that could be very useful? For example to categorise per addon / app?

orta commented:
I've always asked (or [automated asking](http://danger.systems)) contributors to add [CHANGELOG entries](https://github.com/danger/danger-js/blob/master/changelog.md) from a user's perspective. Trying to automate it from PRs and commits has never worked for me as their audience is different (targeted at devs)

tmeasday commented:
I think making PR titles more "user focused" is ok even if commits remain
developer focused?
On Sat, 20 May 2017 at 12:03 am, Orta <notifications@github.com> wrote:

> I've always asked (or automated asking <http://danger.systems>)
> contributors to add CHANGELOG entries
> <https://github.com/danger/danger-js/blob/master/changelog.md> from a
> user's perspective. Trying to automate it from PRs and commits has never
> worked for me as their audience is different (targeted at devs)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/storybooks/storybook/pull/1047#issuecomment-302711257>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIFygRipsvWH-Njfm2J67cmxIiK_TbUks5r7aETgaJpZM4NeEj8>
> .
>

shilman commented:
Nice @orta , love the concept of Danger CI. That's really cool.

I'm going to propose that we run with manual release summaries and edited PR titles:
- For major releases like 3.0.0, the manual summary might take some work and link to blog posts etc., and there is a large backlog of noisy PRs that are all grouped and collapsed.
- For minor releases, the release may only consist of a list of the merged PRs. It's up to the person doing the release how they want to edit the CHANGELOG.

The rationale is that policing things at the commit level is too much overhead, but moving forward we should be able to enforce sensibly-named PRs that are user-centric.

It's an experiment, and we can fall back to the tried-and-true method of manually editing the CHANGELOG after a few releases if it doesn't prove out.

shilman committed:
Merge branch 'master' into 1046-3.0-release-notes

shilman commented:
Best practices for CHANGELOGS for anybody who's interested: http://keepachangelog.com/en/0.3.0/

shilman commented:
@ndelangen latest version with the more granular grouping:

<details>
<summary>
Bug Fixes
</summary>

- #53 Left menu goes blank when switching items #991
- #1016 Restore storybook-addon-links #1021
- #1042 Update README #1043
- #1052 [3.0.0alpha] build-storybook no longer supports relative paths #1058

</details>
<details>
<summary>
Other
</summary>

- #733 Issue triage guidelines #1024

</details>
<details>
<summary>
Maintenance
</summary>

- #759 Deprecation warnings & discontinued addon support for React 16 #1008
- #773 Change npm organisation from kadira to storybook #996

</details>
<details>
<summary>
Open Issues
</summary>

- #953 Typescript compatibility #989

</details>
<details>
<summary>
Addons
</summary>

- #1017 Deprecate addon-links and addon-actions as "built-in addons" #1038,#1025

</details>
<details>
<summary>
PRs without issues
</summary>

- #1074 Update link to Storyshots addon
- #1065 Activating Open Collective
- #1064 Removed `>` typo in Info prop table header
- #1063 chore(jest): update to v20's default configuration file
- #1062 Upgraded webpack PostCSS plugin definition
- #1049 fix(codemod): update for new package names
- #1044 MOVE & RENAME and storyshots && CHANGE “React Storybook” > “Storybook”
- #1030 Added component description from __docgenInfo, Added the ability to get the PropType from __docgenInfo for prop table, Fixed an issue with PropVal rendering an object
- #1026 Switch back to non-fork of react-inspector
- #1001 Fixes for types for knobs `number`
- #998 Update jest to the latest version 🚀
- #992 Update links in PRs section of CONTRIBUTING.md
- #988 Updated dead getstorybook.io links to storybooks.js.org
- #982 First pass at update the contribution instructions
- #980 Fix linting in addon-info
- #977 Fix typo in README
- #974 Fixing link
- #973 Do not ignore story.js in prepublish

</details>

ndelangen reviewed (changes requested):
We need to change the CHANGELOG in the root

shilman committed:
Instructions for generating release notes (design only, not yet implemented)

shilman commented:
@ndelangen i moved the CHANGELOG to the root. let me know if you think i should make a copy instead.

<img width="350" alt="3_0_release_notes_by_shilman_ _pull_request__1047_ _storybooks_storybook" src="https://cloud.githubusercontent.com/assets/488689/26276122/6d6a4904-3db4-11e7-84ee-21073be123a4.png">

shilman committed:
Updated release steps

ndelangen committed:
Merge branch 'master' into 1046-3.0-release-notes

shilman committed:
Merge branch 'storybooks/master' into 1046-3.0-release-notes

# Conflicts:
#	HISTORY.md
#	ROADMAP.md

shilman committed:
Update release notes using latest pr-log

shilman committed:
Merge remote-tracking branch 'origin/1046-3.0-release-notes' into 1046-3.0-release-notes

shilman committed:
Update CONTRIBUTING notes, CHANGELOG format, skip "cleanup" labels

ndelangen committed:
Merge branch 'master' into 1046-3.0-release-notes

shilman committed:
Updated changelog format to be compatible with github-release-from-changelog

shilman committed:
Update CHANGELOG date format

shilman committed:
PR "misc" label => "other" for CHANGELOG
